lms/views/decorators/h_api.py

```diff
@@ -122,3 +124,124 @@ def wrapper(request, jwt):  # pylint: disable=too-many-branches
+        # auto provisioning features for.
+        try:
+            oauth_consumer_key = request.params["oauth_consumer_key"]
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "oauth_consumer_key" missing from LTI params')
+        enabled_consumer_keys = request.registry.settings["auto_provisioning"]
+        if oauth_consumer_key not in enabled_consumer_keys:
+            return wrapped(request, jwt)
+
+        try:
+            tool_consumer_instance_guid = request.params["tool_consumer_instance_guid"]
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "tool_consumer_instance_guid" missing from LTI params')
+
+        try:
+            context_id = request.params["context_id"]
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "context_id" missing from LTI params')
+
+        # Show the user an error page if the group hasn't been created yet and
+        # the user isn't allowed to create the group because they aren't an
+        # administrator, instructor or teaching assistant.
+        try:
+            roles = request.params["roles"].lower()
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "roles" missing from LTI params')
+        if not any(role in roles for role in ("administrator", "instructor", "teachingassisstant")):
+            raise HTTPBadRequest('Teacher must configure assignment first')
+
+        group = models.Group.get(request.db, tool_consumer_instance_guid, context_id)
+
+        if group:
+            # The group has already been created in h, nothing more to do here.
+            return wrapped(request, jwt)
+
+        group = models.Group(
+            tool_consumer_instance_guid=tool_consumer_instance_guid,
+            context_id=context_id,
+        )
+
+        # The URL of h's create group API endpoint.
+        create_group_api_url = request.registry.settings["h_api_url"] + "/groups"
+
+        # Our OAuth 2.0 client_id and client_secret for authenticating to the h API.
+        client_id = request.registry.settings["h_client_id"]
+        client_secret = request.registry.settings["h_client_secret"]
+
+        # Generate the name for the new group.
+        try:
+            name = util.generate_group_name(request.params)
+        except MissingContextTitleError:
+            raise HTTPBadRequest('Required parameter "context_title" missing from LTI params')
+
+        authority = request.registry.settings["h_authority"]
+        username = util.generate_username(request.params)
+
+        # Create the group in h.
+        try:
+            response = requests.post(
+                create_group_api_url,
+                auth=(client_id, client_secret),
+                data=json.dumps({"name": name}),
+                headers={
+                    "X-Forwarded-User": "acct:{username}@{authority}".format(username=username, authority=authority),
```

Comment:
Are there any encoding risks here?

Reply:
The actual string formatting `"acct:{username}@{authority}".format(username=username, authority=authority)` will be fine because all three of these are unicode strings (since this is Python 3).

What's less good is that `requests.post()` crashes if a header value contains a non-Latin-1 character. This is nothing to do with our `lms` app, just the normal behaviour of `requests` in Python 3:

```python
>>> requests.post("http://example.com", headers={"X-Forwarded-User": "Snowman ☃"})
...
UnicodeEncodeError: 'latin-1' codec can't encode character '\u2603' in position 8: ordinal not in range(256)
```

I think this may actually be correct behaviours from `requests` - are non-ASCII characters allowed in HTTP headers (according to the HTTP spec, and to common implementations of HTTP)? I'm not sure without researching it but I think the answer is at least that it's not a good idea to put them in. And I'm not sure why `requests` (or more likely part of the Python stdlib that `requests` is calling) is trying to encode with Latin-1 instead of ASCII here.

_Can_ either of `username` or `authority` here contain non-ASCII chars?

`username`: No. In the LMS app usernames are SHA1's so ASCII and in h usernames generally only [a subset of ASCII chars](https://github.com/hypothesis/h/blob/37cd304261134cf1658f497e236e01f1c85b97f9/h/models/user.py#L17) are allowed.

`authority`: In the case of the LMS app specifically I think no. `authority` here comes from the LMS app deployment variable `H_AUTHORITY` which is `"lms.hypothes.is"`. I'm sure there's nothing preventing you from setting `H_AUTHORITY` to ` ☃☃☃⛄☃☃☃` but we're not going to do that.

So in the case of the LMS app in particular I don't think there's any way for non-ASCII chara to slip in here?

I also wondered whether we needed to change the API generally, such as changing the value of `X-Forwarded-User` to be a base64-encoded userid or something, so that userids with non-ASCII chars could be used:

* `username`: I don't think we have to worry about, since h usernames can't contain non-ASCII chars anyway.
* `authority`: I haven't looked, but I don't think there's any validation that these don't contain non-ASCII chars, but the only authorities that're valid are ones that we add to the production DB (e.g. that we add an authclient for) so we can probably safely assume that they're ASCII?

It might be interesting to see what would happen if you called h's user create API with non-ASCII chars in the `X-Forwarded-User`, if you can get an HTTP client to let you do that.

Reply:
Teacher->Instructor change done.